content/blog/2022/zero-units.md

```diff
@@ -0,0 +1,348 @@
+---
+title: Not All Zeros are Equal
+sub: And every 'best practice' comes with caveats
+author: miriam
+date: 2022-08-04
+image:
+  src: blog/2022/slip-hazard.jpg
+  alt: |
+    Black text on bright yellow sign,
+    Caution, slip hazard,
+    with stick figure falling backwards.
+tags:
+  - Article
+  - CSS
+summary: |
+  There's a well-established 'best practice'
+  that CSS authors (and linters and minifiers)
+  should remove units from any `0` value.
+  It's a fine rule in most cases,
+  but there are a few common situations
+  where it will break your code.
+---
+
+{% import 'embed.macros.njk' as embed %}
+
+I've been working on a redesign of my personal site,
+and found I was fixing the same issue over and over.
+Make the change, test it, commit, and then...
+why is it broken again?
+
+## The intended behavior
+
+When setting typography in a design,
+I like to 'outdent' lists --
+pulling the list markers
+(bullets or numbers)
+out into the margin of the document,
+so that the list contents align
+with the content on either side.
+
+If you're reading this
+on the OddBird website
+with a wide enough browser,
+you'll see that we do that here:
+
+- On wide screens, the bullets of this list should be hanging in the margin.
+- But on narrow browsers, there's not enough 'margin' available.
+- For the list-markers to remain visible on small screens,
+  we switch to the browser-default
+  'hanging indent' --
+  list markers aligned to surrounding content,
+  and list contents indented.
+
+There are various ways we could handle that indent/outdent logic
+(and container queries could be useful).
+For my site,
+I decided to set up an `--outdent` custom property
+on typesetting containers.
+The `--outdent` variable conveys if/when and how much
+margin is available for content.
+
+- By default, for small screens, the `--outdent` is `0`.
+- When a container has more space,
+  I toggle the `--outdent` to something like `-1em`.
+
+Some elements (like figures)
+get the outdent applied directly to a margin:
+
+```css
+figure {
+  margin-inline-start: var(--outdent);
+}
+```
+
+But the list logic is a bit more complicated.
+Since list markers hang 'outside the list' by default,
+we need `0` list-padding on large screens
+and additional padding on small screens
+to achieve an indent.
+I do that with a `calc()` function:
+
+```css
+ol, ul {
+  /* (1em + 0 == 1em) and (1em + -1em == 0) */
+  padding-inline-start: calc(1em + var(--outdent));
+}
+
+li {
+  /* nested lists should not outdent */
```

Comment:
Is this comment correct? should it be a nested list instead of an li?

Reply:
ah, I see... I noticed I was setting the --outdent value in the ol, ul which is why when I added this here, it didn't do anything (event with the unit). When I moved the outdent value to the `html {}` then the `.ol, .ul` didn't override this li anymore.

Reply:
oh right, you also need the outer setting - I can add that to the demo. I use `li` to change it for nested lists - since anything nested would be inside an `li`.